File: 141. Linked List Cycle/main.c
#include <limits.h>
#include <stdbool.h>
#include <stdio.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
bool hasCycle(struct ListNode *head) {
    struct ListNode *slow = head;
    struct ListNode *fast = head;

    while (fast != NULL && fast->next != NULL) {
        slow = slow->next;
        fast = fast->next->next;

        if (slow == fast) {
            return true;
        }
    }

    return false;
}

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
bool hasCycle(struct ListNode *head) {
    while (head != NULL) {
        if (head->val == INT_MAX) {
            return true;
        }

        head->val = INT_MAX;
        head = head->next;
    }

    return false;
}

#define ARRAY_SIZE 10000

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
bool hasCycle(struct ListNode *head) {
    struct ListNode **array = (struct ListNode **)calloc(ARRAY_SIZE, sizeof(struct ListNode *));
    if (array == NULL) {
        return false;
    }

    while (head != NULL) {
        size_t i = 0;
        while (array[i] != NULL) {
            if (head == array[i]) {
                free(array);
                return true;
            }
            i++;
        }

        array[i] = head;
        head = head->next;
    }

    free(array);
    return false;
}

File: 141. Linked List Cycle/main.py
import sys
from collections import defaultdict
from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        while head is not None:
            if head.val == sys.maxsize:
                return True
            head.val = sys.maxsize
            head = head.next

        return False


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        slow = head
        fast = head

        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True

        return False


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        nodes = set()

        while head is not None:
            if head in nodes:
                return True
            nodes.add(head)
            head = head.next

        return False


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        nodes = dict()

        while head is not None:
            if head in nodes:
                return True
            nodes[head] = 1
            head = head.next

        return False


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        nodes = defaultdict(int)

        while head is not None:
            if head in nodes:
                return True
            nodes[head] = 1
            head = head.next

        return False


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        nodes = defaultdict(int)

        while head is not None:
            if nodes[head] != 0:
                return True
            nodes[head] = 1
            head = head.next

        return False


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        marker = ListNode()

        while head is not None:
            if head == marker:
                return True
            next_node = head.next
            head.next = marker
            head = next_node

        return False
